Hash support implementation

github-sajan opened the pull request.

github-sajan committed:
P4C-DPDK - Add support for Hash extern

github-sajan committed:
Hash Extern Implementation

github-sajan committed:
Add support for HASH extern

github-sajan committed:
P4C-DPDK : Add support for Hash extern

usha1830 commented on backends/dpdk/dpdkHelpers.cpp:
```diff
@@ -159,10 +159,94 @@ bool ConvertStatementToDpdk::preorder(const IR::AssignmentStatement *a) {
                     auto field = (*e->expr->arguments)[0];
-                    i = new IR::DpdkGetHashStatement(e->object->getName(),
-                                                     field->expression, left);
+                    /* Check if the Hash Parameters belong to same header/metadata structures
```
Here we can state the requirement of having the hash parameters come from same structure and from consecutive fields. Instead of what is being checked, why it is being checked would more sense.

github-sajan replied:
Addressed

usha1830 commented on backends/dpdk/dpdkHelpers.cpp:
```diff
@@ -159,10 +159,94 @@ bool ConvertStatementToDpdk::preorder(const IR::AssignmentStatement *a) {
+                    */
+                    if (auto b = base->expression->to<IR::Expression>()) {
+                        if (!b->is<IR::Constant>())
+                            ::error("Expecting const expression '%1%' for 'base' value in"
```
Please add error type. Uou can take a look at the available error types in lib/err_catalog.h
Here you can use, ERR_UNEXPECTED

github-sajan replied:
Addressed

usha1830 commented on backends/dpdk/dpdkHelpers.cpp:
```diff
@@ -159,10 +159,94 @@ bool ConvertStatementToDpdk::preorder(const IR::AssignmentStatement *a) {
+                            maxValue = b->to<IR::Constant>()->asUnsigned();
+                            // Check whether max value is power of 2 or not
+                            if (maxValue == 0 || ((maxValue & (maxValue - 1)) != 0)) {
+                                ::error("Invalid Max value '%1%'. DPDK Target expect 'Max'"
```
Likewise.

github-sajan replied:
Addressed

usha1830 commented on backends/dpdk/dpdkHelpers.cpp:
```diff
@@ -346,6 +430,203 @@ bool ConvertStatementToDpdk::preorder(const IR::AssignmentStatement *a) {
+                                               IR::Vector<IR::Expression>& components) {
+    if (auto params = field->expression->to<IR::Member>()) {
+        auto typeInfo = typemap->getType(params);
+        if (typeInfo->is<IR::Type_Header>()) {
```
2 lines can be combined into 1
if (auto typeheader = typeInfo->to<IR::Type_Header>())

github-sajan replied:
Addressed

usha1830 commented on backends/dpdk/dpdkHelpers.cpp:
```diff
@@ -346,6 +430,203 @@ bool ConvertStatementToDpdk::preorder(const IR::AssignmentStatement *a) {
+            if (auto params = field1->expression->to<IR::Member>()) {
+                auto typeInfo = typemap->getType(params);
+                if (typeInfo->is<IR::Type_Header>()) {
+                    auto typeheader = typeInfo->to<IR::Type_Header>();
```
Likewise.

github-sajan replied:
Addressed

usha1830 commented on backends/dpdk/dpdkHelpers.cpp:
```diff
@@ -346,6 +430,203 @@ bool ConvertStatementToDpdk::preorder(const IR::AssignmentStatement *a) {
+}
+
+/* This function processes the hash parameters and update the fields in metadata structure
+   with modified field name, generates mov instruction for each field and store the fields
```
This comment is not very clear.
We can say " This function moves the hash parameter fields to metadata structure if they happen to come from different header/metadata structures or are not consecutively laid in the structure they are coming from" or something on the similar lines

github-sajan replied:
Addressed

usha1830 commented on backends/dpdk/dpdkHelpers.cpp:
```diff
@@ -346,6 +430,203 @@ bool ConvertStatementToDpdk::preorder(const IR::AssignmentStatement *a) {
+            for (auto it : typeheader->fields) {
+                IR::ID name(params->member.name + "." + it->name);
+                IR::ID fldName(refmap->newName(typeheader->name.toString() + "_" + it->name));
+                auto rt = new IR::Member(new IR::PathExpression(IR::ID("h")), name);
```
Why we need to create the header field as well? This should already be available, right?
I think only metadata field name should be created

github-sajan replied:
This code is redundant. Will remove it

usha1830 commented on backends/dpdk/dpdkHelpers.cpp:
```diff
@@ -346,6 +430,203 @@ bool ConvertStatementToDpdk::preorder(const IR::AssignmentStatement *a) {
+/* This function returns the header/metadata structure name */
+cstring ConvertStatementToDpdk::getHdrMdStrName(const IR::Member* mem) {
+    cstring sName = "";
+    if ((mem != nullptr) && (mem->expr != nullptr) &&
```
The first 3 conditions are common and also the last assignment statement as well.
Can this be refactored to return when those 3 conditions are not met?

usha1830 reviewed (changes requested):
Looks fine for functionality. 
Please check comments for some suggestions on code refactoring

github-sajan committed:
P4C-DPDK - Add support for Hash extern
This PR is to support Hash Extern
Hash(PSA_HashAlgorithm_t algo)

Following prototype of 'get_hash' method of Hash extern is supported

1. Compute the hash for data.
   O get_hash<D>(in D data);

2. Compute the hash for data, with modulo by max, then add base.
   O get_hash<T, D>(in T base, in D data, in T max);
   @return (base + (h % max)) where h is the hash value.

   Here, Implementation is limited to constants/const expression
   for 'base' and 'max' and value for 'max' is supposed
   to be powers of 2 eg. 16, 32 or 256.

e.g :
   Hash<bit<16>>(PSA_HashAlgorithm_t.CRC16) h;

   1. b.data = h.get_hash({hdr.ethernet.dstAddr, hdr.ethernet.srcAddr, hdr.ethernet.etherType});
   2. b.data = h.get_hash(base, {hdr.ethernet.srcAddr, hdr.ethernet.etherType}, max_val);
   3. b.data = h.get_hash({hdr.ethernet.srcAddr, hdr.ipv4.protocol});

In implementation, if get_hash parameters are consecutive fields and belong to same headers/metadata
then first and last fields are used in hash instruction. But if parameters belong to
different headers/metadata, then those parameters are populated into user metadata consecutively
and then first and last fields among those are used in hash instruction

github-sajan committed:
This PR is to support Hash Extern
Hash(PSA_HashAlgorithm_t algo)

Following prototype of 'get_hash' method of Hash extern is supported

1. Compute the hash for data.
   O get_hash<D>(in D data);

2. Compute the hash for data, with modulo by max, then add base.
   O get_hash<T, D>(in T base, in D data, in T max);
   @return (base + (h % max)) where h is the hash value.

   Here, Implementation is limited to constants/const expression
   for 'base' and 'max' and value for 'max' is supposed
   to be powers of 2 eg. 16, 32 or 256.

e.g :
   Hash<bit<16>>(PSA_HashAlgorithm_t.CRC16) h;

   1. b.data = h.get_hash({hdr.ethernet.dstAddr, hdr.ethernet.srcAddr, hdr.ethernet.etherType});
   2. b.data = h.get_hash(base, {hdr.ethernet.srcAddr, hdr.ethernet.etherType}, max_val);
   3. b.data = h.get_hash({hdr.ethernet.srcAddr, hdr.ipv4.protocol});

In implementation, if get_hash parameters belong to same headers/metadata and are contigous fields
then first and last fields are used for hash value computation in the instruction. But if parameters belong to
different headers/metadata, then those parameters are populated into user metadata contiguously
and then first and last fields among those are used for hash value computation in the instruction

usha1830 commented on backends/dpdk/dpdkHelpers.cpp:
```diff
@@ -194,9 +194,14 @@ bool ConvertStatementToDpdk::preorder(const IR::AssignmentStatement *a) {
-                    /* Check if the Hash Parameters belong to same header/metadata structures
-                       or if the parameters are consecutive fields if belong to same 
-                       header/metadata structures
+                    /* All the Hash parameters should be in contiguos memory for hash
```
contiguous*

github-sajan replied:
Addressed

usha1830 commented on backends/dpdk/dpdkHelpers.cpp:
```diff
@@ -217,24 +222,35 @@ bool ConvertStatementToDpdk::preorder(const IR::AssignmentStatement *a) {
                         }
                     }
 
+                    /* All the hash parameters should be in contiguos memory for hash
```
comment not needed in both the places. It is the same comment, so should be ok to be present in one place

github-sajan replied:
Addressed

usha1830 commented on backends/dpdk/dpdkHelpers.cpp:
```diff
@@ -580,8 +561,8 @@ bool ConvertStatementToDpdk::checkIfBelongToSameHdrMdStructure(const IR::Argumen
-/* This function processes Hash parameters and checks if all parameters
-   are consecutive field or not in a header/metadata structure.
+/* This function processes hash parameters and checks if all parameters
+   are contiguos field or not in a header/metadata structure.
```
contiguous*

github-sajan replied:
Addressed

github-sajan committed:
This PR is to support Hash Extern
Hash(PSA_HashAlgorithm_t algo)

Following prototype of 'get_hash' method of Hash extern is supported

1. Compute the hash for data.
   O get_hash<D>(in D data);

2. Compute the hash for data, with modulo by max, then add base.
   O get_hash<T, D>(in T base, in D data, in T max);
   @return (base + (h % max)) where h is the hash value.

   Here, Implementation is limited to constants/const expression
   for 'base' and 'max' and value for 'max' is supposed
   to be powers of 2 eg. 16, 32 or 256.

e.g :
   Hash<bit<16>>(PSA_HashAlgorithm_t.CRC16) h;

   1. b.data = h.get_hash({hdr.ethernet.dstAddr, hdr.ethernet.srcAddr, hdr.ethernet.etherType});
   2. b.data = h.get_hash(base, {hdr.ethernet.srcAddr, hdr.ethernet.etherType}, max_val);
   3. b.data = h.get_hash({hdr.ethernet.srcAddr, hdr.ipv4.protocol});

In implementation, if get_hash parameters belong to same headers/metadata and are contigous fields
then first and last fields are used for hash value computation in the instruction. But if parameters belong to
different headers/metadata, then those parameters are populated into user metadata contiguously
and then first and last fields among those are used for hash value computation in the instruction

psivanup reviewed (approved):
Looks ok to me.

usha1830 reviewed (approved):
Looks fine
